subprojects/robotpy-wpilib/wpilib/timedrobotpy.py

```diff
@@ -0,0 +1,396 @@
+_kFramework_Timed = tInstances.kFramework_Timed
+
+microsecondsAsInt = int
+
+
+class _Callback:
+
+    def __init__(
+        self,
+        func: Callable[[], None],
+        periodUs: microsecondsAsInt,
+        expirationUs: microsecondsAsInt,
+    ) -> None:
+        self.func = func
+        self._periodUs = periodUs
+        self.expirationUs = expirationUs
+
+    @classmethod
+    def makeCallBack(
+        cls,
+        func: Callable[[], None],
+        startTimeUs: microsecondsAsInt,
+        periodUs: microsecondsAsInt,
+        offsetUs: microsecondsAsInt,
+        currentTimeUs: microsecondsAsInt,
+    ) -> "_Callback":
+
+        callback = _Callback(func=func, periodUs=periodUs, expirationUs=startTimeUs)
+
+        callback.expirationUs = offsetUs + callback.calcFutureExpirationUs(
+            currentTimeUs
+        )
+        return callback
+
+    def calcFutureExpirationUs(
+        self, currentTimeUs: microsecondsAsInt
+    ) -> microsecondsAsInt:
+        # increment the expiration time by the number of full periods it's behind
+        # plus one to avoid rapid repeat fires from a large loop overrun.
+        #
+        # This routine is called when either:
+        #   this callback has never ran and self.expirationUs is
+        #   TimedRobot._starttimeUs and we are calculating where the first
+        #   expiration should be relative to TimedRobot._starttimeUs
+        # or:
+        #   this call back has ran and self.expirationUs is when it was scheduled to
+        #   expire and we are calculating when the next expirations should be relative
+        #   to the last scheduled expiration
+        #
+        # We assume currentTime ≥ self.expirationUs rather than checking for it since the
+        # callback wouldn't be running otherwise.
+        #
+        # We take when we previously expired or when we started: self.expirationUs
+        # add + self._periodUs to get at least one period in the future
+        # then calculate how many whole periods we are behind:
+        #   ((currentTimeUs - self.expirationUs) // self._periodUs)
+        # and multiply that by self._periodUs to calculate how much time in full
+        # periods we need to skip to catch up, and add that to the sum to calculate
+        # when we should run again.
+        return (
+            self.expirationUs
+            + self._periodUs
+            + ((currentTimeUs - self.expirationUs) // self._periodUs) * self._periodUs
+        )
+
+    def setOrderedFieldNextExpirationUs(self, currentTimeUs: microsecondsAsInt) -> None:
+        self.expirationUs = self.calcFutureExpirationUs(currentTimeUs)
+
+    def __lt__(self, other) -> bool:
+        return self.expirationUs < other.expirationUs
+
+    def __bool__(self) -> bool:
+        return True
+
+    def __repr__(self) -> str:
+        return f"{{func={self.func.__name__}, _periodUs={self._periodUs}, expirationUs={self.expirationUs}}}"
```

Comment:
This should probably just be a dataclass, rather than reimplementing `__repr__` by hand?